clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -653,6 +694,85 @@ public void onFailure(RuntimeException e) {
         }
     }
 
+    /**
+     * For each partition which needs validation, make an asynchronous request to get the end-offsets for the partition
+     * with the epoch less than or equal to the epoch the partition last saw.
+     *
+     * Requests are grouped by Node for efficiency.
+     */
+    private void validateOffsetsAsync() {
+        Set<TopicPartition> partitions = subscriptions.partitionsNeedingValidation(time.milliseconds());
+        Map<TopicPartition, SubscriptionState.FetchPosition> positionMap = partitions.stream()
+                .collect(Collectors.toMap(Function.identity(), subscriptions::position));
+
+        final Map<Node, Map<TopicPartition, SubscriptionState.FetchPosition>> regrouped =
+                regroupPartitionMapByNode(positionMap);
+
+        regrouped.forEach((node, dataMap) -> {
+            subscriptions.setNextAllowedRetry(dataMap.keySet(), time.milliseconds() + requestTimeoutMs);
+
+            final Map<TopicPartition, Metadata.LeaderAndEpoch> cachedLeaderAndEpochs = positionMap.keySet()
+                    .stream()
+                    .collect(Collectors.toMap(Function.identity(), metadata::leaderAndEpoch));
+
+            RequestFuture<OffsetsForLeaderEpochClient.OffsetForEpochResult> future = offsetsForLeaderEpochClient.sendAsyncRequest(node, positionMap);
+            future.addListener(new RequestFutureListener<OffsetsForLeaderEpochClient.OffsetForEpochResult>() {
+                @Override
+                public void onSuccess(OffsetsForLeaderEpochClient.OffsetForEpochResult offsetsResult) {
+                    Map<TopicPartition, OffsetAndMetadata> truncationWithoutResetPolicy = new HashMap<>();
+                    if (!offsetsResult.partitionsToRetry().isEmpty()) {
+                        subscriptions.setNextAllowedRetry(offsetsResult.partitionsToRetry(), time.milliseconds() + retryBackoffMs);
+                        metadata.requestUpdate();
+                    }
+
+                    // For each OffsetsForLeader response, check if the end-offset is lower than our current offset
+                    // for the partition. If so, it means we have experienced log truncation and need to reposition
+                    // that partition's offset.
+                    offsetsResult.endOffsets().forEach((respTopicPartition, respEndOffset) -> {
+                        Metadata.LeaderAndEpoch currentLeader = metadata.leaderAndEpoch(respTopicPartition);
```

Comment:
I would think we'd want to use the epoch in `FetchPosition.currentLeader` for this validation, but maybe we're ensured that they are in sync?